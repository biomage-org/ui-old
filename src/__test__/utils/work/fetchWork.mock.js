const mockCacheKeyMappings = {
  A: 'fd3161a878f67ebf54018720cffd6a66', // pragma: allowlist secret
  B: '10250a11679234110a1c260d6fd81d3c', // pragma: allowlist secret
  C: '76bf160685c4c80c67abd9a701da23e6', // pragma: allowlist secret
  D: '21671038b9ac73c1f08a94c8213cb872', // pragma: allowlist secret
  E: '33faa711a94a2028b5bae1778126aec0', // pragma: allowlist secret
};

const mockGenesListData = ['A', 'B', 'C', 'D'];

const mockGeneExpressionData = {
  A: {
    rawExpression: {
      min: 0,
      max: 6.8,
      mean: 1.68,
      stdev: 2.597331964,
      expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
    },
    truncatedExpression: {
      min: 0,
      max: 6.8,
      mean: 1.68,
      stdev: 2.597331964,
      expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
    },
  },
  B: {
    rawExpression: {
      min: 0,
      max: 6.8,
      mean: 1.702857143,
      stdev: 2.551115536,
      expression: [0, 0, 0, 2.56, 0, 6.8, 2.56],
    },
    truncatedExpression: {
      min: 0,
      max: 6.8,
      mean: 1.702857143,
      stdev: 2.551115536,
      expression: [0, 0, 0, 2.56, 0, 6.8, 2.56],
    },
  },
  C: {
    rawExpression: {
      min: 0,
      max: 3.4,
      mean: 1.68,
      stdev: 2.141525936,
      expression: [0, 0, 0, 3.56, 0, 4.8, 3.4],
    },
    truncatedExpression: {
      min: 0,
      max: 3.4,
      mean: 1.68,
      stdev: 2.141525936,
      expression: [0, 0, 0, 3.56, 0, 4.8, 3.4],
    },
  },
  D: {
    rawExpression: {
      min: 0,
      max: 6.8,
      mean: 1.68,
      stdev: 2.597331964,
      expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
    },
    truncatedExpression: {
      min: 0,
      max: 6.8,
      mean: 1.68,
      stdev: 2.597331964,
      expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
    },
  },
  E: { hello: 'world' },
};

const mockGeneExpressionDataZScore = {
  D: {
    zScore: [
      -0.6468175894669735,
      -0.6468175894669735,
      -0.6468175894669735,
      -0.04620125639049807,
      -0.6468175894669735,
      1.971253605994586,
      0.6622180082638063,
    ],
  },
};

const mockCacheGet = jest.fn((x) => {
  const mockCacheContents = {
    fd3161a878f67ebf54018720cffd6a66: 'A', // pragma: allowlist secret
    '10250a11679234110a1c260d6fd81d3c': 'B', // pragma: allowlist secret
    f5c957411a28de68f35e1f5c8a29da7e: 'C', // pragma: allowlist secret
    '33faa711a94a2028b5bae1778126aec0': 'E', // pragma: allowlist secret
  };

  if (x in mockCacheContents) {
    return mockGeneExpressionData[mockCacheContents[x]];
  }

  return null;
});
const mockCacheSet = jest.fn();
const mockCacheRemove = jest.fn();

const mockDispatchWorkRequest = jest.fn(() => true);

const mockSeekFromS3 = jest.fn();

const mockCacheModule = {
  get: jest.fn((x) => mockCacheGet(x)),
  set: jest.fn((key, val) => mockCacheSet(key, val)),
  _remove: jest.fn((key) => mockCacheRemove(key)),
};

const mockSeekWorkResponseModule = {
  __esModule: true,
  seekFromS3: mockSeekFromS3,
  dispatchWorkRequest: mockDispatchWorkRequest,
};

const mockQcPipelineStartDate = '2021-01-01T01:01:01.000Z';

const mockReduxState = (experimentId, environment = 'testing') => () => ({
  networkResources: {
    environment,
  },
  backendStatus: {
    [experimentId]: {
      status: {
        pipeline: {
          status: 'SUCCEEDED',
          startDate: mockQcPipelineStartDate,
        },
      },
    },
  },
});

export {
  mockGeneExpressionData, mockGeneExpressionDataZScore,
  mockGenesListData,
  mockCacheKeyMappings,
  mockCacheGet, mockCacheSet,
  mockDispatchWorkRequest, mockSeekFromS3,
  mockReduxState, mockQcPipelineStartDate,
  mockCacheModule, mockSeekWorkResponseModule,
};
