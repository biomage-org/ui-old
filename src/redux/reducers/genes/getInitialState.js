import ExpressionMatrix from 'utils/ExpressionMatrix/ExpressionMatrix';

const initialViewState = {
  fetching: false,
  error: false,
  data: [],
};

const getInitialState = () => ({
  properties: {
    loading: [],
    views: {},
    data: {},
  },
  expression: {
    loading: [],
    error: false,
    matrix: new ExpressionMatrix(),
    downsampledMatrix: new ExpressionMatrix(),
    downsampledCellOrder: [],
    views: {},
  },
  selected: [],
  focused: undefined,
  markers: {
    loading: false,
    error: false,
    ETag: null,
  },
});

export { initialViewState };
export default getInitialState;
