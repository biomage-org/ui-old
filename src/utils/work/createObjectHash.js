import { MD5 } from 'object-hash';

const createObjectHash = (object) => MD5(object);

export default createObjectHash;
