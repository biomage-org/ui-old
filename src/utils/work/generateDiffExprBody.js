import { getCellSetKey } from 'utils/cellSets';

const generateDiffExprBody = (experimentId, comparisonGroup, comparisonType, extras) => ({
  name: 'DifferentialExpression',
  experimentId,
  cellSet: getCellSetKey(comparisonGroup.cellSet),
  compareWith: getCellSetKey(comparisonGroup.compareWith),
  basis: getCellSetKey(comparisonGroup.basis),
  comparisonType,
  ...extras,
});

export default generateDiffExprBody;
