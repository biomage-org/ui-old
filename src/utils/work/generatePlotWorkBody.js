import { plotTypes } from 'utils/constants';

const composeDotPlotWorkBody = (config) => {
  const [filterGroup, filterKey] = config.selectedPoints.split('/');

  return {
    name: plotTypes.DOT_PLOT,
    useMarkerGenes: config.useMarkerGenes,
    numberOfMarkers: config.nMarkerGenes,
    customGenesList: config.selectedGenes,
    groupBy: config.selectedCellSet,
    filterBy: {
      group: filterGroup,
      key: filterKey || 'All',
    },
  };
};

const composeImgPlotWorkBody = (config) => ({
  name: 'GetImgPlot',
  type: 'ridgePlot',
  features: config.shownGene,
  plotSubType: config.PlotSubType,
});

const generatePlotWorkBody = (plotType, config) => {
  switch (plotType) {
    case plotTypes.DOT_PLOT:
      return composeDotPlotWorkBody(config);
    case plotTypes.IMG_PLOT:
      return composeImgPlotWorkBody(config);
    default: {
      throw new Error('Plot type doesn\'t exist');
    }
  }
};

export default generatePlotWorkBody;
