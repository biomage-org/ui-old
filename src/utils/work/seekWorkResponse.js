import dayjs from 'dayjs';

import getAuthJWT from 'utils/getAuthJWT';
import fetchAPI from 'utils/http/fetchAPI';
import unpackResult from 'utils/work/unpackResult';
import parseResult from 'utils/work/parseResult';
import WorkTimeoutError from 'utils/errors/http/WorkTimeoutError';
import WorkResponseError from 'utils/errors/http/WorkResponseError';
import httpStatusCodes from 'utils/http/httpStatusCodes';

const throwResponseError = (response) => {
  throw new Error(`Error ${response.status}: ${response.text}`, { cause: response });
};

// getRemainingWorkerStartTime returns how many more seconds the worker is expected to
// need to be running with an extra 1 minute for a bit of leeway
const getRemainingWorkerStartTime = (creationTimestamp) => {
  const now = new Date();
  const creationTime = new Date(creationTimestamp);
  const elapsed = parseInt((now - creationTime) / (1000), 10); // gives second difference

  // we assume a worker takes up to 5 minutes to start
  const totalStartup = 5 * 60;
  const remainingTime = totalStartup - elapsed;
  // add an extra minute just in case
  return remainingTime + 60;
};

const seekFromS3 = async (ETag, experimentId, taskName) => {
  let response;
  try {
    const url = `/v2/workResults/${experimentId}/${ETag}`;

    response = await fetchAPI(url);
  } catch (e) {
    if (e.statusCode === httpStatusCodes.NOT_FOUND) {
      return null;
    }

    throw e;
  }

  const storageResp = await fetch(response.signedUrl);

  if (!storageResp.ok) {
    throwResponseError(storageResp);
  }

  const unpackedResult = await unpackResult(storageResp);
  const parsedResult = parseResult(unpackedResult, taskName);

  return parsedResult;
};

const dispatchWorkRequest = async (
  experimentId,
  body,
  timeout,
  ETag,
  requestProps,
) => {
  console.error('dispatching work request', body);
  const { default: connectionPromise } = await import('utils/socketConnection');
  const io = await connectionPromise;

  const timeoutDate = dayjs().add(timeout, 's').toISOString();
  const authJWT = await getAuthJWT();

  const request = {
    ETag,
    socketId: io.id,
    experimentId,
    ...(authJWT && { Authorization: `Bearer ${authJWT}` }),
    timeout: timeoutDate,
    body,
    ...requestProps,
  };

  const timeoutPromise = new Promise((resolve, reject) => {
    const id = setTimeout(() => {
      reject(new WorkTimeoutError(timeoutDate, request));
    }, timeout * 1000);

    io.on(`WorkerInfo-${experimentId}`, (res) => {
      const { response: { podInfo: { name, creationTimestamp, phase } } } = res;

      const extraTime = getRemainingWorkerStartTime(creationTimestamp);
      if (phase === 'Pending' && extraTime > 0) {
        console.log(`worker ${name} started at ${creationTimestamp}. Adding ${extraTime} seconds to timeout.`);
        clearTimeout(id);
        setTimeout(() => {
          reject(new WorkTimeoutError(timeoutDate, request));
        }, (timeout + extraTime) * 1000);
      }
    });
  });

  const responsePromise = new Promise((resolve, reject) => {
    io.on(`WorkResponse-${ETag}`, (res) => {
      const { response } = res;

      if (response.error) {
        const { errorCode, userMessage } = response;
        console.error(errorCode, userMessage);

        return reject(
          new WorkResponseError(errorCode, userMessage, request),
        );
      }

      // If no error, the response should be ready on S3.
      // In this case, return true
      return resolve();
    });
  });

  const result = Promise.race([timeoutPromise, responsePromise]);

  // TODO switch to using normal WorkRequest for v2 requests
  io.emit('WorkRequest-v2', request);

  return result;
};

export { dispatchWorkRequest, seekFromS3 };
