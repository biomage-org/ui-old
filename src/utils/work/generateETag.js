import { Environment } from 'utils/deploymentInfo';
import config from 'config';
import getExtraDependencies from 'utils/work/getExtraDependencies';
import createObjectHash from './createObjectHash';

// Disable unique keys to reallow reuse of work results in development
const DISABLE_UNIQUE_KEYS = [
  'GetEmbedding',
];

const generateETag = async (
  experimentId,
  body,
  extras,
  qcPipelineStartDate,
  environment,
  dispatch,
  getState,
) => {
  // If caching is disabled, we add an additional randomized key to the hash so we never reuse
  // past results.
  let cacheUniquenessKey = null;

  if (
    environment !== Environment.PRODUCTION
    && localStorage.getItem('disableCache') === 'true'
    && !DISABLE_UNIQUE_KEYS.includes(body.name)
  ) {
    cacheUniquenessKey = Math.random();
  }

  const extraDependencies = await getExtraDependencies(
    experimentId, body, dispatch, getState,
  );

  let ETagBody;

  // They `body` key to create ETAg for gene expression is different
  // from the others, causing the generated ETag to be different
  if (body.name === 'GeneExpression') {
    ETagBody = {
      experimentId,
      missingGenesBody: body,
      qcPipelineStartDate,
      extras,
      cacheUniquenessKey,
      workerVersion: config.workerVersion,
      extraDependencies,
    };
  } else {
    ETagBody = {
      experimentId,
      body,
      qcPipelineStartDate,
      extras,
      cacheUniquenessKey,
      workerVersion: config.workerVersion,
      extraDependencies,
    };
  }

  return createObjectHash(ETagBody);
};

export default generateETag;
