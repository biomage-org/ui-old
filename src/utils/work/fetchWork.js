/* eslint-disable no-underscore-dangle */
import { MD5 } from 'object-hash';

import config from 'config';

import { Environment, isBrowser } from 'utils/deploymentInfo';
import { getBackendStatus } from 'redux/selectors';

import cache from 'utils/cache';
import { dispatchWorkRequest, seekFromS3 } from 'utils/work/seekWorkResponse';

const createObjectHash = (object) => MD5(object);

// Disable unique keys to reallow reuse of work results in development
const DISABLE_UNIQUE_KEYS = [
  'GetEmbedding',
];

const generateETag = (
  experimentId,
  body,
  extras,
  qcPipelineStartDate,
  environment,
) => {
  // If caching is disabled, we add an additional randomized key to the hash so we never reuse
  // past results.
  let cacheUniquenessKey = null;

  if (
    environment !== Environment.PRODUCTION
    && localStorage.getItem('disableCache') === 'true'
    && !DISABLE_UNIQUE_KEYS.includes(body.name)
  ) {
    cacheUniquenessKey = Math.random();
  }

  let ETagBody = {
    experimentId,
    body,
    qcPipelineStartDate,
    extras,
    cacheUniquenessKey,
    workerVersion: config.workerVersion,
  };

  // They `body` key to create ETAg for gene expression is different
  // from the others, causing the generated ETag to be different
  if (body.name === 'GeneExpression') {
    ETagBody = {
      experimentId,
      missingGenesBody: body,
      qcPipelineStartDate,
      extras,
      cacheUniquenessKey,
      workerVersion: config.workerVersion,
    };
  }

  return createObjectHash(ETagBody);
};

// const decomposeBody = async (body, experimentId) => {
//   const { genes: requestedGenes } = body;
//   const missingDataKeys = {};
//   const cachedData = {};

//   await Promise.all(requestedGenes.map(async (g) => {
//     const newBody = {
//       ...body,
//       genes: g,
//     };

//     const key = createObjectHash({ experimentId, newBody });
//     const data = await cache.get(key);

//     if (data) {
//       cachedData[g] = data;
//     } else {
//       missingDataKeys[g] = key;
//     }
//   }));

//   return { missingDataKeys, cachedData };
// };

// Temporarily using gene expression without local cache
const fetchGeneExpressionWorkWithoutLocalCache = async (
  experimentId,
  timeout,
  body,
  backendStatus,
  environment,
  broadcast,
  extras,
) => {
  // If new genes are needed, construct payload, try S3 for results,
  // and send out to worker if there's a miss.
  const { pipeline: { startDate: qcPipelineStartDate } } = backendStatus;

  const ETag = generateETag(
    experimentId,
    body,
    extras,
    qcPipelineStartDate,
    environment,
  );

  // Then, we may be able to find this in S3.
  const response = await seekFromS3(ETag, experimentId);

  if (response) return response;

  // If there is no response in S3, dispatch workRequest via the worker
  try {
    await dispatchWorkRequest(
      experimentId,
      body,
      timeout,
      ETag,
      {
        ETagPipelineRun: qcPipelineStartDate,
        broadcast,
        ...extras,
      },
    );
  } catch (error) {
    console.error('Error dispatching work request: ', error);
    throw error;
  }

  return await seekFromS3(ETag, experimentId);
};

// const fetchGeneExpressionWork = async (
//   experimentId,
//   timeout,
//   body,
//   backendStatus,
//   environment,
//   broadcast,
//   extras,
// ) => {
//   // Get only genes that are not already found in local storage.
//   const { missingDataKeys, cachedData } = await decomposeBody(body, experimentId);

//   const missingGenes = Object.keys(missingDataKeys);

//   if (missingGenes.length === 0) {
//     return cachedData;
//   }

//   // If new genes are needed, construct payload, try S3 for results,
//   // and send out to worker if there's a miss.
//   const { pipeline: { startDate: qcPipelineStartDate } } = backendStatus;

//   const missingGenesBody = { ...body, genes: missingGenes };

//   const ETag = generateETag(
//     experimentId,
//     missingGenesBody,
//     extras,
//     qcPipelineStartDate,
//     environment,
//   );

//   // Then, we may be able to find this in S3.
//   let response = await seekFromS3(ETag, experimentId);

//   // If there is no response in S3, dispatch workRequest via the worker
//   if (!response) {
//     try {
//       await dispatchWorkRequest(
//         experimentId,
//         missingGenesBody,
//         timeout,
//         ETag,
//         {
//           ETagPipelineRun: qcPipelineStartDate,
//           broadcast,
//           ...extras,
//         },
//       );
//     } catch (error) {
//       console.error('Error dispatching work request: ', error);
//       throw error;
//     }
//   }

//   response = await seekFromS3(ETag, experimentId);

//   Object.keys(missingDataKeys).forEach(async (gene) => {
//     await cache.set(missingDataKeys[gene], response[gene]);
//   });

//   return response;
// };

const fetchWork = async (
  experimentId,
  body,
  getState,
  optionals = {},
) => {
  const {
    extras = undefined, timeout = 180, broadcast = false, fetchS3Data = true,
  } = optionals;
  const backendStatus = getBackendStatus(experimentId)(getState()).status;

  const { environment } = getState().networkResources;
  if (!isBrowser) {
    throw new Error('Disabling network interaction on server');
  }

  if (environment === Environment.DEVELOPMENT && !localStorage.getItem('disableCache')) {
    localStorage.setItem('disableCache', 'true');
  }

  const { pipeline: { startDate: qcPipelineStartDate } } = backendStatus;
  if (body.name === 'GeneExpression') {
    return fetchGeneExpressionWorkWithoutLocalCache(
      experimentId, timeout, body, backendStatus, environment, broadcast, extras,
    );
  }

  const ETag = generateETag(
    experimentId,
    body,
    extras,
    qcPipelineStartDate,
    environment,
  );

  // First, let's try to fetch this information from the local cache.
  const data = await cache.get(ETag);

  if (data) {
    return data;
  }

  console.log('BEFORE SEEK');
  // Then, we may be able to find this in S3.
  let response = await seekFromS3(ETag, experimentId, fetchS3Data);

  if (response) return response;

  // If there is no response in S3, dispatch workRequest via the worker
  try {
    await dispatchWorkRequest(
      experimentId,
      body,
      timeout,
      ETag,
      {
        PipelineRunETag: qcPipelineStartDate,
        broadcast,
        ...extras,
      },
    );

    response = await seekFromS3(ETag, experimentId, fetchS3Data);
  } catch (error) {
    console.error('Error dispatching work request', error);
    throw error;
  }

  // If a work response is in s3, it is cacheable
  // (the cacheable or not option is managed in the worker)
  await cache.set(ETag, response);
  return response;
};

export {
  fetchWork,
  generateETag,
};
