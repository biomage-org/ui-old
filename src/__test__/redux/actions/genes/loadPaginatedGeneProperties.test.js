import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import loadPaginatedGeneProperties from 'redux/actions/genes/loadPaginatedGeneProperties';
import getInitialState from 'redux/reducers/genes/getInitialState';

import { dispatchWorkRequest, seekFromS3 } from 'utils/work/seekWorkResponse';

import {
  GENES_PROPERTIES_LOADING,
  GENES_PROPERTIES_LOADED_PAGINATED,
  GENES_PROPERTIES_ERROR,
} from 'redux/actionTypes/genes';

jest.mock('utils/work/seekWorkResponse', () => ({
  __esModule: true, // this property makes it work
  dispatchWorkRequest: jest.fn(() => true),
  seekFromS3: jest.fn(() => new Promise((resolve) => { resolve(null); })),
}));

jest.mock('utils/getTimeoutForWorkerTask', () => ({
  __esModule: true, // this property makes it work
  default: () => 60,
}));

const mockStore = configureStore([thunk]);

const experimentId = '1234';
const properties = ['a', 'b', 'c'];
const componentUuid = 'asd';

const backendStatus = {
  [experimentId]: {
    status: {
      pipeline: {
        status: 'SUCCEEDED',
        startDate: '2021-01-01T01:01:01.000Z',
      },
    },
  },
};

describe('loadPaginatedGeneProperties action', () => {
  beforeEach(async () => {
    jest.clearAllMocks();

    seekFromS3
      .mockReset()
      .mockImplementation(() => null);
  });

  it('Does not dispatch when some of the properties are already loading', async () => {
    const initialState = getInitialState();
    const store = mockStore({
      backendStatus,
      genes:
      {
        ...initialState,
        properties: {
          ...initialState.properties,
          loading: ['b'],
        },
      },
    });

    store.dispatch(loadPaginatedGeneProperties(experimentId, properties, componentUuid, {}));
    expect(store.getActions().length).toEqual(0);
  });

  it('Dispatches appropriately on success condition', async () => {
    const initialState = getInitialState();
    const store = mockStore({
      genes:
      {
        ...initialState,
      },
      backendStatus,
      networkResources: {
        environment: 'testing',
      },
    });

    seekFromS3
      .mockReset()
      .mockImplementationOnce(() => null)
      .mockImplementation(() => Promise.resolve({
        total: 2,
        rows: [
          {
            gene_names: 'a',
            dispersions: 1,
          },
          {
            gene_names: 'b',
            dispersions: 1,
          },
        ],
      }));

    const tableState = {
      sorter: {
        field: 'gene_names',
        order: 'ascend',
      },
      pagination: {
        current: 2,
        pageSize: 20,
      },
    };

    await store.dispatch(
      loadPaginatedGeneProperties(experimentId, properties, componentUuid, tableState),
    );

    expect(store.getActions().length).toEqual(2);

    const loadingAction = store.getActions()[0];
    expect(loadingAction).toMatchSnapshot();
    expect(loadingAction.type).toEqual(GENES_PROPERTIES_LOADING);

    const loadedAction = store.getActions()[1];
    expect(loadedAction).toMatchSnapshot();
    expect(loadedAction.type).toEqual(GENES_PROPERTIES_LOADED_PAGINATED);

    const dispatchParams = dispatchWorkRequest.mock.calls[0];
    expect(dispatchParams).toMatchSnapshot();
  });

  it('Dispatches appropriately on error condition', async () => {
    const store = mockStore({
      genes:
      {
        ...getInitialState(),
      },
      backendStatus,
      networkResources: {
        environment: 'testing',
      },
    });

    seekFromS3
      .mockReset()
      .mockImplementationOnce(() => null)
      .mockImplementation(() => Promise.reject(new Error('random error!')));

    const tableState = {
      sorter: {
        field: 'gene_names',
        order: 'ascend',
      },
      pagination: {
        current: 2,
        pageSize: 20,
      },
    };

    await store.dispatch(
      loadPaginatedGeneProperties(experimentId, properties, componentUuid, tableState),
    );

    expect(store.getActions().length).toEqual(2);

    const loadingAction = store.getActions()[0];
    expect(loadingAction.type).toEqual(GENES_PROPERTIES_LOADING);
    expect(loadingAction).toMatchSnapshot();

    const loadedAction = store.getActions()[1];
    expect(loadedAction.type).toEqual(GENES_PROPERTIES_ERROR);
    expect(loadedAction).toMatchSnapshot();
  });
});
