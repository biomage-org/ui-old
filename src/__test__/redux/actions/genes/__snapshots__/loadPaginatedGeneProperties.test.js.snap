// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadPaginatedGeneProperties action Dispatches appropriately on error condition 1`] = `
{
  "payload": {
    "componentUuid": "asd",
    "properties": [
      "a",
      "b",
      "c",
    ],
  },
  "type": "genes/propertiesLoading",
}
`;

exports[`loadPaginatedGeneProperties action Dispatches appropriately on error condition 2`] = `
{
  "payload": {
    "componentUuid": "asd",
    "error": [Error: random error!],
    "experimentId": "1234",
    "properties": [
      "a",
      "b",
      "c",
    ],
  },
  "type": "genes/propertiesError",
}
`;

exports[`loadPaginatedGeneProperties action Dispatches appropriately on success condition 1`] = `
{
  "payload": {
    "componentUuid": "asd",
    "properties": [
      "a",
      "b",
      "c",
    ],
  },
  "type": "genes/propertiesLoading",
}
`;

exports[`loadPaginatedGeneProperties action Dispatches appropriately on success condition 2`] = `
{
  "payload": {
    "componentUuid": "asd",
    "data": {
      "a": {
        "dispersions": 1,
      },
      "b": {
        "dispersions": 1,
      },
    },
    "experimentId": "1234",
    "properties": [
      "a",
      "b",
      "c",
    ],
    "total": 2,
  },
  "type": "genes/propertiesLoadedPaginated",
}
`;

exports[`loadPaginatedGeneProperties action Dispatches appropriately on success condition 3`] = `
[
  "1234",
  {
    "limit": 20,
    "name": "ListGenes",
    "offset": 20,
    "orderBy": "gene_names",
    "orderDirection": "ASC",
    "selectFields": [
      "gene_names",
      "a",
      "b",
      "c",
    ],
  },
  60,
  "f7aeae492eecc9d6ea35ab2cc0212ad1",
  {
    "PipelineRunETag": "2021-01-01T01:01:01.000Z",
    "broadcast": false,
  },
]
`;
